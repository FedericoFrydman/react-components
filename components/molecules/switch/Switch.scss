@import '../../assets/styles/colors.scss';

.lsux-switch {
  background-color: rgba(0, 0, 0, 0);
  border: none;
  cursor: pointer;
  display: inline-block;
  position: relative;
  -ms-transition: background-color 800ms ease-in-out;
  -webkit-transition: 0.4s;
  -webkit-transition: background-color 800ms ease-in-out;
  transition: 0.4s;
  transition: background-color 800ms ease-in-out;
  vertical-align: top;

  &:focus {
    outline: none;
    .lsux-switch__knob {
      box-shadow: 0 0 0 1.5px var(--switch--enabled-checked-background);
    }
  }

  &--small {
    height: 16px;
    width: 28px;
  }
  &--large {
    height: 24px;
    width: 42px;
  }

  &.disabled {
    cursor: not-allowed;
  }
  .lsux-switch__slider {
    background-color: var(--switch--enabled-unchecked-background);
    border-color: $N350;
    border-radius: 48px;
    border-style: solid;
    border-width: 0px;
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    -ms-transition: background-color 800ms ease-in-out;
    -webkit-transition: background-color 800ms ease-in-out;
    transition: background-color 800ms ease-in-out;
  }
  .lsux-switch__knob {
    background-color: $N00;
    border-color: $N350;
    border-radius: 50%;
    border-style: solid;
    border-width: 0.5px;
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.24);
    position: absolute;
    -webkit-transition: transform 0.4s;
    transition: transform 0.4s;
    &--small {
      height: 14px;
      left: calc(50% - 8px / 2 - 9px);
      top: calc(50% - 14px / 2);
      width: 14px;
    }
    &--large {
      height: 22px;
      left: calc(50% - 22px / 2 - 9px);
      top: calc(50% - 22px / 2);
      width: 22px;
    }
  }
  &.checked {
    .lsux-switch__knob {
      background-color: $N00;
      border-color: $N350;

      &--small {
        -ms-transform: translateX(12px);
        -webkit-transform: translateX(12px);
        transform: translateX(12px);
      }
      &--large {
        -ms-transform: translateX(18px);
        -webkit-transform: translateX(18px);
        transform: translateX(18px);
      }
    }
    .lsux-switch__slider {
      background-color: var(--switch--enabled-checked-background);
      border-color: var(--switch--enabled-checked-background);
      -webkit-transition: border-color 750ms ease-in-out;
      -ms-transition: border-color 750ms ease-in-out;
      transition: border-color 750ms ease-in-out;
    }
  }
  &.disabled {
    .lsux-switch__slider {
      background-color: var(--switch--disabled-unchecked-background);
      border-color: var(--switch--disabled-unchecked-background);
      -webkit-transition: border-color 800ms ease-in-out;
      -ms-transition: border-color 800ms ease-in-out;
      transition: border-color 800ms ease-in-out;
    }
    .lsux-switch__knob {
      background-color: $N00;
      border-color: $N350;
    }
  }
}
