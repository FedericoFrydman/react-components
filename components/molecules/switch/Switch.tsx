import { Ref, HTMLProps, forwardRef } from 'react';
import cn from 'classnames';

import './Switch.scss';

export interface SwitchProps extends Omit<HTMLProps<HTMLButtonElement>, 'size' | 'type' | 'css'> {
  /**
   * Indicates that the switch is checked (turned on)
   */
  isChecked?: boolean;
  /**
   * Indicates that the switch control is currently disabled
   */
  disabled?: boolean;
  /**
   * Indicates the size of the switch control
   */
  switchSize?: 'small' | 'large';
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLButtonElement>;
  /**
   * Action trigger when switch is pressed
   */
  onClick?: (e: React.MouseEvent) => void | unknown;
}

const SwitchComponent = (
  { isChecked = false, onClick, disabled = false, switchSize = 'large', classNames = [], ...props }: SwitchProps,
  ref: SwitchProps['ref'],
) => {
  const classSwitch = 'lsux-switch';
  const cnSwitch = cn(
    classSwitch,
    `${classSwitch}--${switchSize}`,
    {
      checked: isChecked,
      disabled: disabled,
    },
    classNames,
  );

  const classKnob = `${classSwitch}__knob`;
  const cnKnob = cn(classKnob, `${classKnob}--${switchSize}`);
  const ariaLabel = props['aria-label'] || 'Switch';

  return (
    <button {...props} onClick={onClick} className={cnSwitch} ref={ref} disabled={disabled} aria-label={ariaLabel}>
      <span className="lsux-switch__slider"></span>
      <span className={cnKnob}></span>
    </button>
  );
};

/**
 * Switch switchs between an binary action, "on" and "off" state. The response should be instant and requires no button to confirm selection.
 */
export const Switch = forwardRef<HTMLButtonElement, SwitchProps>(SwitchComponent);
/**
 * Toggle backwards compatible.
 */
//export const Toggle = forwardRef(SwitchComponent);
export const Toggle = forwardRef<HTMLButtonElement, SwitchProps>(SwitchComponent);
export type ToggleProps = SwitchProps;
