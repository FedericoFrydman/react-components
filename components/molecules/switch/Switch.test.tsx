import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Switch, SwitchProps } from './Switch';

describe('Switch Component', () => {
  const testId = 'test-switch';
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Switch />);
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the switch the provided classNames', () => {
          const switchElem = renderSwitch({ classNames });

          expect(switchElem).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLButtonElement>();

    it('provides a ref to the switch', () => {
      expect(renderSwitch({ ref })).toEqual(ref.current);
    });
  });

  const renderSwitch = (props: SwitchProps = {}): HTMLElement => {
    render(<Switch {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
