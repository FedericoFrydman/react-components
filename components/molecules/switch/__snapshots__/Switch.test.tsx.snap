// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Component should render the component 1`] = `
<button
  aria-label="Switch"
  className="lsux-switch lsux-switch--large"
  disabled={false}
>
  <span
    className="lsux-switch__slider"
  />
  <span
    className="lsux-switch__knob lsux-switch__knob--large"
  />
</button>
`;
